pinot-core/src/main/java/org/apache/pinot/core/segment/processing/aggregator/DistinctCountCPCSketchAggregator.java

```diff
@@ -30,7 +31,7 @@ public DistinctCountCPCSketchAggregator() {
   }
 
   @Override
-  public Object aggregate(Object value1, Object value2) {
+  public Object aggregate(Object value1, Object value2, Map<String, String> functionParameters) {
```

Comment:
Would be possible to use jackson's `JsonObject` instead of `Map<String, String>`? We have historically (ab)use of `Map<String, String>` which forces us to do some tricks when complex parameters are needed.

Reply:
Actually this Map<String, String> encapsulates all that the merge rollup has to do and currently we dont have any other way to pass in params. Instead of propagating the map itself, can we extract out what we need from the map and create a separate Object/class with well defined types for the AggregatorContext. It becomes clear whats needed for the AggregatorContext, with well defined types instead of passing the entire map?  @davecromberge

Reply:
@swaminathanmanish is your idea to change the signature as follows:
```
public Object aggregate(Object value1, Object value2, Object functionParameters)
```
And then each `ValueAggregator` subtype casts this Object to a concrete and more specific type?  To my knowledge, there is no other general purpose type that can be used across all ValueAggregators since each ValueAggregator will have its own function parameters.

It would help to flesh out your idea a little more so that I can try it out.

Reply:
@davecromberge - Not sure if it needs to be an Object. Instead of passing Map<String, String> to the interface, can we have the class for FunctionParams and individual params as typed fields in that class. We will extract out these fields in getAggregationFunctionParameters. If Im missing something, please let me know. 
If we are going to use these params conditionally with no defaults then we dont have a choice, other than to use Map. 

Class AggregateFunctionParams {
int _sketchNominalEntries; // This will be default or custom value
...}

AggregateFunctionParams will be created in  getAggregationFunctionParameters. Will be have a Map<String, AggregateFunctionParams> 

Object aggregate(Object value1, Object value2, AggregateFunctionParams aggregateFunctionParams) {
int sketchNominalEntries = aggregateFunctionParams._sketchNominalEntries;
...
}

Reply:
Thanks for fleshing out your idea further @swaminathanmanish.   I agree that using a concrete type with given fields would generally lead to more readable and maintainable code.  My biggest concern is that not all value aggregators would work off the same function parameters and thus we might need more flexibility.  Since we are not able to create a unified AggregationFunctionParams type, we would need to use something generic like Map<String, String>, JsonObject or Object which, while providing flexibility does reduce readability.

For example, the theta sketch value aggregator might expect `_sketchNominalEntries` to be set whereas the CPC or HLL function may expect a different parameter which cannot be unified with that of Theta.

Reply:
Alternatively, we might instantiate each value aggregator with the relevant parameters but this would need to be baked into the constructors.  For example, in the `ValueAggregatorFactory`:

```
public static ValueAggregator getValueAggregator(AggregationFunctionType aggregationType, DataType dataType) {
    Map<String, String> params = getAggregationFunctionParameters;

    switch (aggregationType) {
      case MIN:
        return new MinValueAggregator(dataType);
      case MAX:
        return new MaxValueAggregator(dataType);
      case SUM:
        return new SumValueAggregator(dataType);
      case DISTINCTCOUNTHLL:
      case DISTINCTCOUNTRAWHLL:
        return new DistinctCountHLLAggregator();
      case DISTINCTCOUNTTHETASKETCH:
      case DISTINCTCOUNTRAWTHETASKETCH:
        return new DistinctCountThetaSketchAggregator(Integer.parseInt(params.get("_sketchNominalEntries")));
      case DISTINCTCOUNTTUPLESKETCH:
      case DISTINCTCOUNTRAWINTEGERSUMTUPLESKETCH:
      case SUMVALUESINTEGERSUMTUPLESKETCH:
      case AVGVALUEINTEGERSUMTUPLESKETCH:
        return new IntegerTupleSketchAggregator(IntegerSummary.Mode.Sum);
      case DISTINCTCOUNTCPCSKETCH:
      case DISTINCTCOUNTRAWCPCSKETCH:
        return new DistinctCountCPCSketchAggregator();
      case DISTINCTCOUNTULL:
      case DISTINCTCOUNTRAWULL:
        return new DistinctCountULLAggregator();
      default:
        throw new IllegalStateException("Unsupported aggregation type: " + aggregationType);
    }

```

Reply:
Thanks @davecromberge . So we need the flexibility. We can keep it as Map<String, String> then.